rs/messaging/src/routing/stream_handler/tests.rs

```diff
@@ -3468,281 +2870,599 @@ fn process_stream_slices_canister_migration_in_both_subnets_success() {
-        )
-        .build();
+/// Returns a reference to a request in a stream at `stream_index`.
+///
+/// Panics if no such request exists.
+fn request_ref_in_stream(opt_stream: Option<&Stream>, stream_index: u64) -> &Request {
+    match opt_stream.and_then(|stream| stream.messages().get(stream_index.into())) {
+        Some(RequestOrResponse::Request(request)) => request,
+        _ => unreachable!(),
+    }
+}
 
-    let messages = slice
-        .messages()
-        .cloned()
-        .unwrap_or_else(|| StreamIndexedQueue::with_begin(msg_begin));
+/// Returns a reference to a request in a stream slice at `stream_index`.
+///
+/// Panics if no such request exists.
+fn request_ref_in_slice(opt_slice: Option<&StreamSlice>, stream_index: u64) -> &Request {
+    match opt_slice.and_then(|slice| {
+        slice
+            .messages()
+            .and_then(|msgs| msgs.get(stream_index.into()))
+    }) {
+        Some(RequestOrResponse::Request(request)) => request,
+        _ => unreachable!(),
+    }
+}
 
-    if let Some(reject_signals) = signal_config.reject_signals {
-        Stream::with_signals(
-            messages,
-            slice.header().signals_end(),
-            reject_signals.into(),
-        )
-    } else {
-        Stream::new(messages, slice.header().signals_end())
+/// Returns a reference to a response in a stream at `stream_index`.
+///
+/// Panics if no such response exists.
+fn response_ref_in_stream(opt_stream: Option<&Stream>, stream_index: u64) -> &Response {
+    match opt_stream.and_then(|stream| stream.messages().get(stream_index.into())) {
+        Some(RequestOrResponse::Response(response)) => response,
+        _ => unreachable!(),
     }
 }
 
-#[derive(Clone)]
-struct StreamConfig {
-    messages_begin: u64,
-    message_count: u64,
-    signals_end: u64,
-    reject_signals: Option<Vec<RejectSignal>>,
+/// Returns a reference to a response in the stream slice at `stream_index`.
+///
+/// Panics if no such response exists.
+fn response_ref_in_slice(opt_slice: Option<&StreamSlice>, stream_index: u64) -> &Response {
+    match opt_slice.and_then(|slice| {
+        slice
+            .messages()
+            .and_then(|msgs| msgs.get(stream_index.into()))
+    }) {
+        Some(RequestOrResponse::Response(response)) => response,
+        _ => unreachable!(),
+    }
 }
 
-fn generate_outgoing_stream(config: StreamConfig) -> Stream {
-    generate_stream(
-        MessageConfig {
-            sender: *LOCAL_CANISTER,
-            receiver: *REMOTE_CANISTER,
-            begin: config.messages_begin,
-            count: config.message_count,
-        },
-        SignalConfig {
-            end: config.signals_end,
-            reject_signals: config.reject_signals,
-        },
-    )
+/// Returns a clone of the message in a stream at `stream_index`.
+///
+/// Panics if no such message exists.
+fn clone_message_from_stream(opt_stream: Option<&Stream>, stream_index: u64) -> RequestOrResponse {
```

Comment:
I find the `clone_message_from_[stream|slice]` functions unnecessary. Why not simply `[request|response|message]_in_[stream|slice]().clone()`? It would also make it unnecessary to have the `_ref_` in the method names above: just `request_in_stream()` / `response_in_slice()`; or even just `request_in()` for the former.

Reply:
The `clone_message_from_()` functions are used where it doesn't matter which kind of message it is exactly. It also returns `RequestOrResponse` and not `&Request` or `&Response`. So what you are suggesting would be actually be `request_in_stream().clone().into()` which bloats the code for no reason IMO. It also puts emphasis on whether it's a request or a response in cases where it doesn't matter. I can dump the `ref` if you want.

I felt a small helper function like this doesn't matter much in a test and it expresses what I want without unnecessary details or `clone()`, `into()`, `unwrap()` and so forth. I am not entirely consistent with the `unwrap()` I guess you could say since I left it on inductions but not when getting messages from streams. I guess the former is far more likely to fail due to conceptual changes so `unwrap()` in place seemed like the better choice.

Reply:
I was more taking issue with the `clone_` prefix (and the function returning a clone). You could instead have `message_in_[stream|slice]()` functions (which would also allow you to drop the `_ref_` from other function names, since all of them return references now) and just write `message_in_stream(...).clone()` instead of `clone_message_in_stream(...)`.

A second concern (and the reason why I only had two helper functions, covering both streams and slices) is that the more magic you have in your test (`clone_message_from_stream()` vs `response_ref_in_slice()`, and so forth) the more you just have to rely on magic and hope you understand what it does. In (one sort of) an ideal world, tests would be written only using the public APIs of the component being tested, no access to the internals, no magic hiding what's going on.

I'm not insisting you go back to `get_request()` / `get_response()` (which I preferred), but I find a function with `clone_` in the name that simply clones whatever else it looks up, a bit unnecessary.

Reply:
I guess I can make all these functions return references and then ditch the `clone`, `ref` in the name and put the `.clone()` behind the function. I still think not having to write `.into()` in lots of places merits this function.

Reply:
It's a double edged sword. Unfortunately `messages` is optional in `StreamSlice` which makes that annoying to deal with using the same function.

Overall it was quite cumbersome reading the original tests because there was always quite a bit of boilerplate in the way. I tried to find a good compromise of 'two much magic' and cutting off details obscuring what the test actually does. Especially in the later more complex tests where it was just generating identical messages for rerouted messages I felt it was a bit hard to follow what it going on. This is why I put all these `response @33` comments everywhere, though with the current framework it's not strictly necessary anymore.

I was also considering a sort of config to generate the expected stream but felt it was making things too stiff and not really any clearer. I guess I could give that it's own closure now, but I don't think it makes it better.